File: chattools/src/main/java/aytackydln/chattools/telegram/dto/models/Update.java
package aytackydln.chattools.telegram.dto.models;

import aytackydln.chattools.telegram.TelegramChat;
import aytackydln.chattools.telegram.TelegramMessage;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.Data;
import lombok.RequiredArgsConstructor;

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
@Data
@RequiredArgsConstructor
public class Update {
    private int updateId;
    private TelegramMessage message;
    private CallbackQuery callbackQuery;
    private ChatDetails myChatMember;

    @JsonIgnore
    public TelegramChat getUser() {
        if (message != null) {
            return message.getUser();
        } else if (callbackQuery != null) {
            return callbackQuery.getFrom();
        } else if (myChatMember != null) {
            return myChatMember.getChat();
        }
        throw new IllegalArgumentException("User could not be found from update");
    }
}

File: chattools/src/main/kotlin/aytackydln/chattools/telegram/dto/models/Update.kt
package aytackydln.chattools.telegram.dto.models

import aytackydln.chattools.telegram.TelegramChat
import aytackydln.chattools.telegram.TelegramMessage
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.databind.annotation.JsonNaming
import lombok.Data
import lombok.RequiredArgsConstructor
import lombok.extern.jackson.Jacksonized

@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
data class Update(
    val updateId: Long,
    val message: TelegramMessage? = null,
    val callbackQuery: CallbackQuery? = null,
    val myChatMember: ChatDetails? = null,
) {

    @get:JsonIgnore
    val user: TelegramChat?
        get() {
            if (message != null) {
                return message.user as TelegramChat?
            } else if (callbackQuery != null) {
                return callbackQuery.from
            } else if (myChatMember != null) {
                return myChatMember.chat
            }
            throw IllegalArgumentException("User could not be found from update")
        }
}
